### PR TITLE
Makes the pool ready after first connection

### PR DESCRIPTION
**Pull request by flucchese**
This patch changes the behavior of the connection pool, making it ready as soon as the first connection is added to it. This prevents clients from timing out when there are many connections to be added.

### CODE REVIEW

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 1]
```diff
@@ -337,30 +289,49 @@ private PoolingOptions options() {
...
-            manager.blockingExecutor().submit(new ConnectionTask(shardId));
-          }
-        }
-        return enqueue(timeout, unit, maxQueueSize, shardId);
-      }
+    int index = 0;
+    int shId = shardId;
+    do {
+    	synchronized(connections[shId]) {
+    		if ((connections[shId].isEmpty()) && (index < connections.length)) {
```
Comment (haaawk) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: Nit: Unnecessary brackets

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 2]
```diff
@@ -337,30 +289,49 @@ private PoolingOptions options() {
...
+    	    	shId = index++;
```
Comment (haaawk) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: Wrong indent

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 3]
```diff
@@ -337,30 +289,49 @@ private PoolingOptions options() {
...
+    		} else {
+    			break;
+    		}
+    	}
+    } while(true);
```
Comment (haaawk) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:298]: Why not while (connections[shId].isEmpty() && index < connections.length) { ... } ?
Even better:
```
if (connections[shId].isEmpty()) {
  do {
    shId = (shId + 1) % connections.length;
  } while (connections[shId].isEmpty() && shId != shardId);
}
```
to prevent overloading connections for shard 0.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 4]
```diff
@@ -337,30 +289,49 @@ private PoolingOptions options() {
...
+    if ((shId != shardId) && (!connections[shId].isEmpty())) {
```
Comment (haaawk) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:299]: Nit: Unnecessary brackets

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 5]
```diff
@@ -396,7 +367,8 @@ private PoolingOptions options() {
           || maxTotalInFlight.compareAndSet(oldMax, totalInFlightCount)) break;
     }
 
-    int connectionCount = connections[shardId].size() + scheduledForCreation[shardId].get();
```
Comment (haaawk) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:399]: Why did you move connections[shardId].size()? It's better to call it here to get the fresh value as other threads may be changing it.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 6]
```diff
@@ -337,30 +289,49 @@ private PoolingOptions options() {
...
-        } else if (scheduledForCreation[shardId].compareAndSet(0, connectionsPerShard)) {
-          for (int i = 0; i < connectionsPerShard; i++) {
-            // We don't respect MAX_SIMULTANEOUS_CREATION here because it's  only to
-            // protect against creating connection in excess of core too quickly
-            manager.blockingExecutor().submit(new ConnectionTask(shardId));
-          }
-        }
-        return enqueue(timeout, unit, maxQueueSize, shardId);
-      }
+    int index = 0;
```
Comment (haaawk) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:288]: This selection of connection to different shard does not seem right. This way we may never try to reconnect to a shard that loses all its connections.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 7]
```diff
@@ -168,117 +166,79 @@ public void run() {
...
+					addCallback(newConnection,
+							handleErrors(newConnection.initAsync(), initExecutor),
+							initFuture,
+							count);
+				}
+			}
+
+			@Override
+			public void onFailure(Throwable t) {
+				int f = failed.incrementAndGet();
```
Comment (haaawk) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: I think we should try to connect again if a connection failed. At least some number of times.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 8]
```diff
@@ -337,30 +289,49 @@ private PoolingOptions options() {
...
-            // protect against creating connection in excess of core too quickly
-            manager.blockingExecutor().submit(new ConnectionTask(shardId));
-          }
-        }
-        return enqueue(timeout, unit, maxQueueSize, shardId);
-      }
+    int index = 0;
+    int shId = shardId;
+    do {
+    	synchronized(connections[shId]) {
```
Comment (haaawk) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: This is wrong. This method is supposed to be non-blocking (it returns future) and this is a synchronous lock.

**Review by haaawk — CHANGES_REQUESTED** [review]
I think this is not the right way to fix the problem. The main issue is that it turns a non-blocking, lock-free code into blocking code that uses locks. 

Another thing is that before we were keeping excessive connections to shards that already have connectionsPerShard connections to increase chances of getting connection to the right shard. You totally dropped it.

Also the code uses 2 spaces indent and you introduced different indentation in various places.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 7 above)
Comment (flucchese) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: The code does that. Notice that the onFailure callback is only invoked when a non-recoverable exception is thrown in the Connection class. In all other cases, the onSuccess is invoked with a closed connection, which we re-open.

Comment (haaawk) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: I know that closed connection is being re-opened but non-recoverable exception does not mean we can't open a new connection instead of this one that failed.

**Comment by haaawk** [comment]
Now that I had another look on the change, I noticed one more important problem. Previously, the code was making sure that we create connectionsPerShard for each shard. The change totally drops this and gives no guarantees whatsoever regarding the distribution of connections among the shards.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 7 above)
Comment (flucchese) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: No, it means exactly that. The exceptions that lead to onFailure are: ClusterNameMismatchException, UnsupportedProtocolVersionException and AuthenticationException. You cannot recover from that. Check the handleErrors() method.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 8 above)
Comment (flucchese) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: The method is non-blocking. This synchronized block protects concurrent access to a non thread-safe structure (ArrayList). It doesn't block the entire method, just this snippet. Concurrent access to non thread-safe structures lead to unpredictable behavior (https://www.amazon.com/Taming-Java-Threads-Allen-Holub/dp/1893115100)

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 7 above)
Comment (haaawk) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: Why are you counting the failures then, instead of terminating the init process after the first non-recoverable exception?

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 1 above)
Comment (flucchese) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: In Java, braces should not be ommitted in one liners. Check this: https://dzone.com/articles/omitting-braces-not-just-a-mat; Also a good reference here: https://www.amazon.com/Effective-Java-Joshua-Bloch/dp/0134685997

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 2 above)
Comment (flucchese) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: Right. My IDE is introducing TABs instead of SPACEs. Need to fix this.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 4 above)
Comment (flucchese) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:299]: See answer above.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 3 above)
Comment (flucchese) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: As for the first suggestion (while...) it's not possible because of the synchronize statement. Regarding the later, that's a good idea, thanks.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 7 above)
Comment (flucchese) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: Because that was the former behavior, which I kept. But you're right, we can fail altogether at the first onFailure execution.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 3 above)
Comment (flucchese) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:298]: As for the first suggestion (while...) it's not possible because of the synchronize statement. Regarding the later, that's a good idea, thanks.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 8 above)
Comment (haaawk) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: Are you serious? 
> It doesn't block the entire method, just this snippet.

What does that even mean?

As an experienced Java architect you surely know that ["All synchronized blocks synchronized on the same object can only have one thread executing inside them at the same time. All other threads attempting to enter the synchronized block **are blocked** until the thread inside the synchronized block exits the block"](http://tutorials.jenkov.com/java-concurrency/synchronized.html)

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 7 above)
Comment (haaawk) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: The former behaviour was that the whole init process of the pool was failing when a single connection got non-recoverable exception. Not waiting for other connections to finish.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 8 above)
Comment (flucchese) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: I think we might not be talking about the same thing. If you don't understand what I meant by that, I strongly advise you to check the reference I posted. The synchronized block is thread safe, not the entire method. Maybe this clarifies the point: https://stackoverflow.com/questions/20906548/why-is-synchronized-block-better-than-synchronized-method

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 1 above)
Comment (haaawk) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: I didn't suggest doing:
```
if ((connections[shId].isEmpty()) && (index < connections.length))
  shId = index++;
```
I suggested doing:
```
if (connections[shId].isEmpty() && index < connections.length) {
  shId = index++;
}
```
No need to put an expression in brackets when it's as simple as connections[shId].isEmpty() or index < connections.length

Comment (flucchese) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: That is irrelevant and I hope we won't spend the company's money debating over curly brackets preferences.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 7 above)
Comment (flucchese) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: I review the contract of Futures.allAsList and it indeed fails whenever a connection fails.  So, as I said, we can fail altogether when at the first onFailure execution.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 6 above)
Comment (flucchese) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:288]: I don't get the point. Can you elaborate on that?

**Review by flucchese — COMMENTED** [review]
> I think this is not the right way to fix the problem. The main issue is that it turns a non-blocking, lock-free code into blocking code that uses locks.
> 
> Another thing is that before we were keeping excessive connections to shards that already have connectionsPerShard connections to increase chances of getting connection to the right shard. You totally dropped it.
> 
> Also the code uses 2 spaces indent and you introduced different indentation in various places.

As notified above, your understanding of concurrency control is deficient. A synchronized block does not block the method or the object. Please review the references I posted in the respective comments.

The identation problem has been fixed.

The number of connections per host continues to be determined by a pooling option (CoreConnectionsPerHost). Hard-coding a literal multiplier (as it was before) is plain wrong.

**Comment by flucchese** [comment]
> Now that I had another look on the change, I noticed one more important problem. Previously, the code was making sure that we create connectionsPerShard for each shard. The change totally drops this and gives no guarantees whatsoever regarding the distribution of connections among the shards.

The code guarantees that each shard has connectionsPerShard connections.

**Comment by flucchese** [comment]
There's a new commit to patch-1 with the modifications that I understood/agreed upon.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 9]
```diff
@@ -168,116 +165,75 @@ public void run() {
...
-                    Lists.newArrayListWithCapacity(needed + Math.max(0, toClose.size() - limit));
-                while (toClose.size() > limit) {
-                  connectionFutures.add(toClose.poll().closeAsync());
-                }
-                final List<Connection> connections =
-                    manager.connectionFactory().newConnections(HostConnectionPool.this, needed);
-                for (Connection connection : connections) {
-                  ListenableFuture<Void> connectionFuture = connection.initAsync();
-                  connectionFutures.add(handleErrors(connectionFuture, initExecutor));
+    final Connection connection,
```
Comment (tarzanek) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:188]: +1 for sticking with old style of whitespace for the sake of review
once we change style, we shall have it in separate PR again for the sake of review

Comment (flucchese) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:188]: I've alreayd committed the fixes for the identation. Should I close this PR and create a new one?

**Comment by flucchese** [comment]
Closing this PR and creating a new one with the formatting fixes.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 6 above)
Comment (haaawk) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:288]: The previous version of the code was making sure that for each shard if there wasn't any connection it would try to create a new one. After this change, the code will just ignore shards that have no connections and instead take connections to different shards.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 8 above)
Comment (haaawk) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: We clearly don't talk about the same thing. The issue I'm raising is that synchronized block is like a mutex. If one thread acquires it, all other threads have to wait (hence block) until the mutex is released. This means a method with a synchronized block is blocking.

**Comment by haaawk** [comment]
> > Now that I had another look on the change, I noticed one more important problem. Previously, the code was making sure that we create connectionsPerShard for each shard. The change totally drops this and gives no guarantees whatsoever regarding the distribution of connections among the shards.
> 
> The code guarantees that each shard has connectionsPerShard connections.



> > I think this is not the right way to fix the problem. The main issue is that it turns a non-blocking, lock-free code into blocking code that uses locks.
> > Another thing is that before we were keeping excessive connections to shards that already have connectionsPerShard connections to increase chances of getting connection to the right shard. You totally dropped it.
> > Also the code uses 2 spaces indent and you introduced different indentation in various places.
> 
> As notified above, your understanding of concurrency control is deficient. A synchronized block does not block the method or the object. Please review the references I posted in the respective comments.
> 
> The identation problem has been fixed.
> 
> The number of connections per host continues to be determined by a pooling option (CoreConnectionsPerHost).

This is not true. There should be CoreConnectionsPerHost created per host after the pool is initialized. This is not guaranteed anymore. It would be fine to have fewer connections after initialization and making sure we keep trying to connect to ensure the right number of connections is established but the current version of the change does not do that.

>Hard-coding a literal multiplier (as it was before) is plain wrong.

What literal multiplier do you mean?

**Comment by haaawk** [comment]
> There's a new commit to patch-1 with the modifications that I understood/agreed upon.

Please incorporate the fixes into the initial commit.

**Comment by haaawk** [comment]
> > Now that I had another look on the change, I noticed one more important problem. Previously, the code was making sure that we create connectionsPerShard for each shard. The change totally drops this and gives no guarantees whatsoever regarding the distribution of connections among the shards.
> 
> The code guarantees that each shard has connectionsPerShard connections.

Could you please explain to me how does the code guarantee that each shard has connectionsPerShard connections? It does guarantee that no shard will have more than connectionsPerShard connections but nothing is guaranteeing that each shard will have exactly connectionsPerShard connections. Moreover, nothing is guaranteeing that each shard will have even a single connection.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 8 above)
Comment (flucchese) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: Yes, that's the definition of a thread-safe block. So, what are you advocating? To avoid using any kind of locks and synchronization mechanisms? To begin with, "connections" should not be an ArrayList.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 6 above)
Comment (flucchese) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:288]: No, the pool will keep opening connections until all shards are full. Review the addCallback method. You're looking at the wrong spot.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 8 above)
Comment (haaawk) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: I'm saying that the code before was thread-safe and lock-free. We should keep it that way. I'm not saying we have to avoid all synchronization mechanisms - only those that block.

Why do you think connections shouldn't be ArrayList?

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 6 above)
Comment (haaawk) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:288]: addCallback is called only on initialization. This method can be called much later than that. Imagine a case when the initialization has successfully finished and then for some reason one of the connections gets dropped. In such a case, borrowConnection should try to recreate such connection. In general, it will but if all connections for a single shard get dropped it won't consider it anymore and will keep using connections to other shards instead.
